src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1056,7 +1057,8 @@ public int getIndex()
         void add(CompactionSSTable sstable)
         {
             this.sstables.add(sstable);
-            this.avg += (sstable.onDiskLength() - avg) / sstables.size();
+            // consider size of all components to reduce chance of out-of-disk
+            this.avg += (sstable.onDiskComponentsSize() - avg) / sstables.size();
```

Comment:
can we have a system property to rollback ?

Reply:
using `onDiskComponentsSize` instead of `onDiskLength` means that we estimate the size of the sstable as the size of not only the data but also the other metadata, SAI, Bloom filters and so on.

We estimate that the result will be as big as the sum of all the components of the input sstables.
I wonder how accurate is this.
Things that come to my mind:
- obsolte Immutable components generations
- bloom filters ? is the bloom filter as large as the sum of the bloom filters for the input sstables
- ... anything else that doesn't grow linearly ?

Maybe we need to add another method to sstable, like "estimatedDisksizeForCompaction" in which we keep only the latest generation of SAI files and we don't count the other components that don't grow linearly ?

Reply:
> obsolte Immutable components generations
bloom filters ? is the bloom filter as large as the sum of the bloom filters for the input sstables
... anything else that doesn't grow linearly ?

Good point. When using `EtcdSSTable#onDiskComponentsSize` (data size + archive size) in CNDB compactor before actually opening sstable readers, older SAI generations in archive file will cause over-estimation.

BF is generally not as large as data file and SAI files, not a big issue. 

> Maybe we need to add another method to sstable, like "estimatedDisksizeForCompaction" in which we keep only the latest generation of SAI files and we don't count the other components that don't grow linearly ?

This is what `SSTableReader#onDiskComponentsSize` is doing, `SSTableReader#components` includes SAI components from current SAI generation.

The issue is in `EtcdSSTable#onDiskComponentsSize` (we use `EtcdSSTable` in compactor to delay sstable opening until actually starting compaction), we only know about the entire archive size.